### PR TITLE
DM-33325: Update measurement uncertainties in meas_extensions_trailedSources

### PR DESCRIPTION
**Pull request by langfzac**
First-pass at adding uncertainties for updated `NaivePlugin` measurements (length, angle, flux).
- Move flux computation to C++ layer, and adds gradient computation with respect to the other `VeresModel` parameters.
- Add `SAFE_CENTROID` flag. (Want the `SdssShape` centroid, so throw flag if we fall back to safe.)
- Move 'non-weighted' length computation to function and compute gradients of both length computations.
- Propagate measurement uncertainties from centroid and second-moments.
- Add tests for gradient computations
- Add monte-carlo tests for length and angle uncertainties. For now, just make sure we're close, or at least over estimate. 

Note: Flux uncertainty underestimates the monte-carlo derived values (often ~50%). This will be addressed in a future PR.

### CODE REVIEW

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 1]
```diff
@@ -226,6 +249,9 @@ def measure(self, measRecord, exposure):
         measRecord.set(self.keyY0Err, y0Err)
         measRecord.set(self.keyX1Err, x0Err)
         measRecord.set(self.keyY1Err, y0Err)
+        measRecord.set(self.keyFluxErr, fluxErr)
+        measRecord.set(self.keyLengthErr, lengthErr)
+        measRecord.set(self.keyAngleErr, thetaErr)
```
Comment (natelust) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py:254]: I can't comment on the individual bits of math and such, but an overall comment is that this is an incredibly long method, that will be hard to follow and maintain by future people. Even with small code comments, its a bit hard to hold the "flow" of the method in your head at once. I am not saying you must do it for this ticket, however, it would be nice if at some point you could factor some of the logical bits that go together into private methods (similar to computeLength, though that is not private).

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 2]
```diff
@@ -296,13 +324,53 @@ def findLength(self, Ixx, Iyy):
         # c_min is given by the case: Ixx == Iyy, ie. a point source.
         # c_max is given by the limit Ixx >> Iyy.
         # Emperically, 0.001 is a suitable lower bound, assuming Ixx > Iyy.
-        z, results = sciOpt.brentq(lambda z: self._computeSecondMomentDiff(z, c),
+        z, results = sciOpt.brentq(lambda z: cls._computeSecondMomentDiff(z, c),
```
Comment (natelust) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py:327]: Are these constants from a math equation you are referencing? Generally try to to use more descriptive variable names rather than single letters. I know this change set is not about that, so I am not going to insist on you refactoring now if you don't want to, but try to keep it in mind next time you touch the code. It will make it much easier to understand to someone else, including future you.

File: src/VeresModel.cc [hunk 3]
```diff
@@ -92,6 +92,43 @@ std::vector<double> VeresModel::gradient(std::vector<double> const& params) cons
     return gradChiSq;
 }
 
+std::tuple<double, std::vector<double>> VeresModel::computeFluxWithGradient(std::vector<double> const& params) const {
+
+    double xc = params[0];     // Centroid x
+    double yc = params[1];     // Centroid y
+    // double flux = params[2];   // Flux (in this case, always 1)
```
Comment (natelust) [inline src/VeresModel.cc:99]: should this be commented out?

File: src/VeresModel.cc [hunk 4]
```diff
@@ -92,6 +92,43 @@ std::vector<double> VeresModel::gradient(std::vector<double> const& params) cons
...
+    double m2 = 0.0;  // sum_i model_i*model_i
+    double md = 0.0;  // sum_i model_i*data_i
+    std::vector<double> gradmd = {0.0,0.0,0.0,0.0,0.0};  // sum_i (gradModel_(i,k)*data_i)
+    std::vector<double> gradmm = {0.0,0.0,0.0,0.0,0.0};  // sum_i (gradModel_(i,k)*model_i)
+    for (int yIndex = 0, yp = _bbox.getBeginY(); yIndex < _bbox.getHeight(); ++yIndex, ++yp) {
+        ImageF::Array::Reference dataRow = _data[yIndex];
+        for (int xIndex = 0, xp = _bbox.getBeginX(); xIndex < _bbox.getWidth(); ++xIndex, ++xp) {
+            double data = dataRow[xIndex];
+            double model = _computeModel(xp,yp,xc,yc,1.0,length,theta);
+            std::array<double, 5> gradModel = _computeGradient(xp,yp,xc,yc,1.0,length,theta);
```
Comment (natelust) [inline src/VeresModel.cc]: spaces after ,

**Review by natelust — APPROVED** [review]

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 1 above)
Comment (langfzac) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py:254]: Fair enough, and yes I think another ticket could tackle this.

File: src/VeresModel.cc (hunk 3 above)
Comment (langfzac) [reply src/VeresModel.cc:99]: Yes. It made sense to me to keep the same function arguments as the other methods, so I just left it as a comment.

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 2 above)
Comment (langfzac) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py:327]: Which variables are you referring to? `z` and `c`?

Comment (natelust) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py:327]: I noticed it here specifically, but in general too, it may seem like overkill often, but it helps someone else get up to speed. It is a bit of a discretion thing, but our dev guide does say when possible prefer descriptive names.

Comment (langfzac) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py:327]: Gotcha. I agree in general. Here, I'm not sure how to do that, since `z` and `c` are transformed variables and don't really have a physical meaning. I just used the same notation I am using in the paper.

Comment (natelust) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py:327]: Ok, that is why I asked, there is certainly a carve out for things like that, commonly you just reference the paper equation describing it in a comment or something like that. Even easier if you put the citation in a Note section in the task docstring, and then the comment can just be something like `# variables c and z as described in eq X`
